src/decoder/plugins/UsfDecoderPlugin.cxx

```diff
@@ -0,0 +1,318 @@
+    }
+    return false;
+}
+
+static int
+usf_tags_target(void * context, const char * name, const char * value)
+{
+    struct UsfTags * tags = ( struct UsfTags * ) context;
+    if (!value) return 0;
+
+    if (tags->setTags) {
+        usf_set_tag("title", name, value, TAG_TITLE, tags) ||
+            usf_set_tag("artist", name, value, TAG_ARTIST, tags) ||
+            usf_set_tag("composer", name, value, TAG_COMPOSER, tags) ||
+            usf_set_tag("game", name, value, TAG_ALBUM, tags) ||
+            usf_set_tag("year", name, value, TAG_DATE, tags) ||
+            usf_set_tag("genre", name, value, TAG_GENRE, tags) ||
+            usf_set_tag("track", name, value, TAG_TRACK, tags);
+    }
+
+    if (strcmp("length", name) == 0) {
+        tags->length = get_length_from_string(value);
+    } else if (strcmp("fade", name) == 0) {
+        tags->fade = get_length_from_string(value);
+    }
+    return 0;
+}
+
+
+inline static double
+track_lengths(double length, double fade)
+{
+    return length + fade;
+
+}
+
+static void
+usf_file_decode(Decoder &decoder, Path path_fs)
+{
+    /* Load the file */
+
+    UsfTags tags;
+    UsfLoaderState state;
+    state.emu = malloc(usf_get_state_size());
+    usf_clear(state.emu);
+    AtScopeExit(state) {
+        free(state.emu);
+    };
+
+    const char* path = path_fs.c_str();
+    const int psf_version_tags = psf_load(path, &stdio_callbacks, 0, 0, 0, usf_tags_target, &tags, 0);
+    const int psf_version_state = psf_load( path, &stdio_callbacks, psf_version_tags, usf_loader, &state, usf_info, &state, 0 );
+
+    if ( psf_version_tags != 0x21 || psf_version_state <= 0 ) {
+        LogWarning(usf_domain, "Error loading usf file");
+        return;
+    }
+
+    usf_set_compare( state.emu, state.enableCompare );
+    usf_set_fifo_full( state.emu, state.enableFIFOFull );
+
+    int32_t sample_rate;
+    usf_render( state.emu, 0, 0, &sample_rate );
+
+    /* initialize the MPD decoder */
+
+    const AudioFormat audio_format(sample_rate, SampleFormat::S16, USF_CHANNELS);
+    assert(audio_format.IsValid());
+
+    double totalLength =  track_lengths(tags.length, tags.fade);
+
+    // Duration
+    decoder_initialized(decoder, audio_format, true, SongTime::FromS(totalLength));
+
+    /* .. and play */
+    DecoderCommand cmd;
+    do {
+        int16_t buf[USF_BUFFER_SAMPLES];
+        const char* result = usf_render(state.emu, buf, USF_BUFFER_FRAMES, 0);
+        if (result != 0) {
+            LogWarning(usf_domain, result);
+            break;
+        }
+
+        // Simple fading
+        double fadeTime = tags.fade;
+        double trackLength = tags.length;
+        if (fadeTime > 0 && totalLength >= 0 && decoder.timestamp > trackLength) {
+            const double vol = 1.0 - ((decoder.timestamp - trackLength) / fadeTime);
+            const double normalizedVol = vol < 0 ? 0 : vol;
+            for (unsigned int i = 0; i < USF_BUFFER_SAMPLES; i++) {
+                buf[i] *= normalizedVol;
```

Comment:
A fading function is necessary (and expected) in a format like this. It is also implemented in the GmeDecoderPlugin, although here it is implemented through a call to the third party library, that applies the fading before returning the samples to GmeDecoderPlugin. This would still be in the decoder, no?
Such a function does not exist in lazyusf, which is why I needed to implement it here.